src/mono/mono/metadata/memory-manager.c

```diff
@@ -49,11 +49,37 @@ unregister_vtable_reflection_type (MonoVTable *vtable)
 		MONO_GC_UNREGISTER_ROOT_IF_MOVING (vtable->type);
 }
 
+// First phase of deletion
+void
+mono_mem_manager_free_objects (MonoMemoryManager *memory_manager)
+{
+	memory_manager->freeing = TRUE;
+
+	g_ptr_array_free (memory_manager->class_vtable_array, TRUE);
+	memory_manager->class_vtable_array = NULL;
+
+	// Must be done before type_hash is freed
+	for (int i = 0; i < memory_manager->class_vtable_array->len; i++)
+		unregister_vtable_reflection_type ((MonoVTable *)g_ptr_array_index (memory_manager->class_vtable_array, i));
+
+	mono_g_hash_table_destroy (memory_manager->type_hash);
+	memory_manager->type_hash = NULL;
+	mono_conc_g_hash_table_foreach (memory_manager->refobject_hash, cleanup_refobject_hash, NULL);
+	mono_conc_g_hash_table_destroy (memory_manager->refobject_hash);
+	memory_manager->refobject_hash = NULL;
+	mono_g_hash_table_destroy (memory_manager->type_init_exception_hash);
+	memory_manager->type_init_exception_hash = NULL;
+}
```

Comment:
memory_manager itself is not freed.

Reply:
It's freed in `mono_mem_manager_free_singleton`.

Reply:
Oh no you're right, oops.

Reply:
It is freed there, my bad.